### PR TITLE
Start DateInput

### PR DESCRIPTION
**Pull request by Ladsgroup**
Bug: [T273002](https://phabricator.wikimedia.org/T273002)

### CODE REVIEW

File: vue-components/src/components/DateInput.vue
Comment (micgro42) [inline vue-components/src/components/DateInput.vue:13]: This also has to reproduce the suffix slot

File: vue-components/src/components/DateInput.vue
Comment (micgro42) [inline vue-components/src/components/DateInput.vue:22]: I think this class should be `wikit-DateInput__CalendarNotice` - it is an element, not a modifier, right?

**Review by micgro42 — COMMENTED** [review]
looks good so far :+1:

File: vue-components/src/components/DateInput.vue
Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue:13]: Done.

File: vue-components/src/components/DateInput.vue
Comment (micgro42) [inline vue-components/src/components/DateInput.vue:9]: InputWithExtender doesn't have a prop "feedback" type. The error prop is good enough

**Review by micgro42 — COMMENTED** [review]
Solid job! just one little detail to fix and it can have my approval :+1:

File: vue-components/src/components/DateInput.vue
Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue]: oh nice. Thanks. done now.

**Comment by SaiSan-WMDE** [comment]
This is really great! 💯 

There's a Reference error that prevents the Basic story to be displayed in Chromatic. Adding some comments here after checking the components in the All story:

- For some reason, the IntroText style is not being applied to the default prompt:
 
Screenshot 2021-03-19 at 12 59 12

- The variants in the All page should have some more vertical space between them (see the All story of the ExtendedNumberInput for reference). I believe that the spacing is achieved by containing the variants in paragraphs.

- In the design specification, it was indicated that text elements should be contained in 24px high "bounding boxes" spaced by 4px. Checking the current implementation, I believe that probably that's not necessary, but that definitely the vertical spacing between elements should be increased to 8px (dimension.spacing.small.value). I wasn't sure where to suggest the change

**Comment by SaiSan-WMDE** [comment]
Charlie made a good point regarding the default active state of the DateInput component that made me think the interaction flow could be improved in the following way:

Current: Once the user focuses on the DateInput, the extender menu opens to display the prompt text.
Desired: Once the user focuses on the DateInput, instead of just the prompt, users also see the resultsIntroText. In the following way:

Screenshot 2021-03-30 at 10 39 48

Since the component is still in the last stages of development, I thought we still are on time to make this improvement 🙏

**Comment by Ladsgroup** [comment]
> This is really great! 100
> 
> There's a Reference error that prevents the _Basic_ story from being displayed in Chromatic. Adding some comments here after checking the components in the _All_ story:
> 
>     * For some reason, the IntroText style is not being applied to the default prompt in Chrome & Safari on MacOs (the style looks good in Firefox/Linux, though):
> 
> 
> Screenshot 2021-03-19 at 12 59 12
> 
>     * The variants in the All page should have some more vertical space between them (see the All story of the ExtendedNumberInput for reference). I believe that the spacing is achieved by containing the variants in paragraphs (which is probably also not ideal).
> 

Added `<br>` in between.

>     * In the design specification, it was indicated that text elements inside the Extender menu should be contained in 24px high "bounding boxes" (aka divs) verticall-spaced by 4px. Checking the current implementation, I'm not sure if that's applied or even if it's necessary, but definitely the vertical spacing between elements should be increased, to probably 8px (dimension.spacing.small.value). I wasn't sure where to suggest the change

Isn't it `$wikit-InputExtender-padding-vertical`?

**Comment by SaiSan-WMDE** [comment]
Thanks for all the changes and fixes, @Ladsgroup! 

> > * For some reason, the IntroText style is not being applied to the default prompt in Chrome & Safari on MacOs (the style looks good in Firefox/Linux, though):

This error still persists: the text-styles applied to the prompt are ignored in MacOS (checked Safari, Firefox and Chrome)

> Isn't it `$wikit-InputExtender-padding-vertical`?

Yes, that token is defining the internal padding of the menu, but I believe that the elements are unaffected because we are using <p>s.

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:23]:
```suggestion
					<Icon type="info-filled" size="small" />
```

Comment (SaiSan-WMDE) [reply vue-components/src/components/DateInput.vue]: my bad! I realized we should be using the icon that accompanies notice messages.

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:23]:
```suggestion
					<Icon type="info" size="small" />
```

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:88]:
```suggestion
		padding-block: $wikit-InputExtender-content-padding-vertical;
		font-family: $font-family-style-label;
```

Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:88]:
```suggestion
		padding-block: $wikit-InputExtender-content-padding-vertical;
		font-family: $font-family-style-label;
```

Comment (SaiSan-WMDE) [reply vue-components/src/components/DateInput.vue:88]: should be added to each one of these, right? 

sorry, I'd usually try this out locally, but I find lots of conflicts that I'm not sure how to resolve when I try to work on this branch locally

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:23]: can we apply `color="subtle"` to the icon or make it inherit the color of the text it accompanies? Any of the options are fine

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:87]: I guess we should add `padding-bottom: $wikit-InputExtender-content-padding-vertical;` here

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:95]: And `padding-bottom: $wikit-InputExtender-content-padding-vertical;` & `padding-top: $wikit-InputExtender-content-padding-vertical;` here

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [inline vue-components/src/components/DateInput.vue:103]: And `padding-top: $wikit-InputExtender-content-padding-vertical;` here

File: vue-components/src/components/DateInput.vue
Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue:95]: Done

File: vue-components/src/components/DateInput.vue
Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue:87]: Done

File: vue-components/src/components/DateInput.vue
Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue:23]: Done

Comment (Ladsgroup) [reply vue-components/src/components/DateInput.vue]: We don't have "subtle" color type. The ones are `'base', 'warning', 'error', 'notice', 'success', 'inherit'`. I change it to inherit.

**Comment by SaiSan-WMDE** [comment]
yay! looks so good! I would approve right away if it weren't (again) for that missing style:

Screenshot 2021-03-30 at 17 58 14

Is there anything we can try there? Should we create a class for it too? The style literally matches that of the input [placeholder](https://www.figma.com/file/gi3JYwSUQYhKS0mHwK25cQwi/WiKit?node-id=7000%3A9060).

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [reply vue-components/src/components/DateInput.vue]: thanks! inherit makes sense. I'm opening a small PR to add those missing colors (I was convinced they were there, but maybe mixed up colors and sizes), but no further change is needed here.

File: vue-components/src/components/DateInput.vue
Comment (micgro42) [inline vue-components/src/components/DateInput.vue:20]: Not blocking this PR: But since we now show the intro text always, do we want the prompt to actually still be a slot? That is where the application can inject whatever markup it wants. Or would we not prefer for this now to be a text-only prop? @SaiSan-WMDE 

Though, this is not blocking the merge and can still be adjusted later on.

**Comment by micgro42** [comment]
This still needs a line in the CHANGELOG.md, I think :)

File: vue-components/src/components/DateInput.vue
Comment (SaiSan-WMDE) [reply vue-components/src/components/DateInput.vue:20]: @Ladsgroup and I just discussed this and originally agreed to keep the slot. But you're right, this is some static text that will always be displayed, so using a text-only prop like in the other cases makes more sense and is more consistent?

Comment (micgro42) [reply vue-components/src/components/DateInput.vue:20]: You tell me 🙃
This really depends on what you envision might be inserted there by the application.
If it is only ever going to be plain text that is always supposed to be styled the same way, then a prop makes more sense.
On the other hand, if you want to keep the possibility open that the application might want to provide some copy that includes a "learn more" link, then we should keep the slot.

If you two discussed this and decided that to keep the slot, then I'm confident that this is the right decision and am very happy to approve this as soon as we have the note in the CHANGELOG.md :)

Comment (SaiSan-WMDE) [reply vue-components/src/components/DateInput.vue:20]: I believe this should be a text slot for now, since I cannot foresee any use cases that justify the use of a slot at the moment. A slot could be introduced later if that changed.

Comment (micgro42) [reply vue-components/src/components/DateInput.vue]: Implemented as requested.

**Review by SaiSan-WMDE — APPROVED** [review]
Just gorgeous. Thanks, Michael and Amir! 💯 

giphy (4)

**Review by micgro42 — APPROVED** [review]

**Pull request merged**
